airbyte-integrations/connectors/source-marketo/source_marketo/source.py

```diff
@@ -0,0 +1,530 @@
+    """
+
+    pass
+
+
+class Lists(IncrementalMarketoStream):
+    """
+    Return list of all lists.
+    API Docs: http://developers.marketo.com/rest-api/endpoint-reference/lead-database-endpoint-reference/#!/Static_Lists/getListsUsingGET
+    """
+
+    pass
+
+
+class MarketoAuthenticator(Oauth2Authenticator):
+    def __init__(self, config):
+        super().__init__(
+            token_refresh_endpoint=f"{config['base_url']}/identity/oauth/token",
+            client_id=config["client_id"],
+            client_secret=config["client_secret"],
+            refresh_token=None,
+        )
+
+    def get_refresh_request_params(self) -> Mapping[str, Any]:
+        payload: MutableMapping[str, Any] = {
+            "grant_type": "client_credentials",
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+
+        return payload
+
+    def refresh_access_token(self) -> Tuple[str, int]:
+        """
+        Returns a tuple of (access_token, token_lifespan_in_seconds)
+        """
+        try:
+            response = requests.request(method="GET", url=self.token_refresh_endpoint, params=self.get_refresh_request_params())
+            response.raise_for_status()
+            response_json = response.json()
+            # print(response_json["access_token"])
+            return response_json["access_token"], response_json["expires_in"]
+        except Exception as e:
+            raise Exception(f"Error while refreshing access token: {e}") from e
+
+
+class SourceMarketo(AbstractSource):
+    """
+    Source Marketo fetch data of personalized multi-channel programs and campaigns to prospects and customers.
+    """
+
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+        """
+        Testing connection availability for the connector by granting the credentials.
+        """
+
+        try:
+            url = f"{config['base_url']}/rest/v1/leads/describe"
+
+            authenticator = MarketoAuthenticator(config)
+
+            session = requests.get(url, headers=authenticator.get_auth_header())
+            session.raise_for_status()
+
+            return True, None
+        except requests.exceptions.RequestException as e:
+            return False, e
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        config["authenticator"] = MarketoAuthenticator(config)
+
+        streams = [ActivityTypes(config), Campaigns(config), Leads(config), Lists(config), Programs(config)]
+
+        for activity in ActivityTypes(config).read_records(sync_mode=None):
+            stream_name = f"activities_{clean_string(activity['name'])}"
+
+            stream_class = type(stream_name, (Activities,), {"activity": activity})
+
+            # instantiate a stream with config
+            stream_instance = stream_class(config)
+            streams.append(stream_instance)
+
+        return streams
```

Comment:
Need more description why the `Activity/_X` streams are built dynamically.
For future improvements: is there any other opportunity we could retrieve all the activity types inside of 1 stream and store them within `activity_type` column in 1 table?

Reply:
Updated.